src/runtime/metadata_module.cc

```diff
@@ -0,0 +1,220 @@
+   * \param symbol The symbol that is being queried.
+   * \return The list of needed NDArray.
+   */
+  Array<NDArray> GetRequiredMetadata(const std::string& symbol) {
+    Array<NDArray> ret;
+    CHECK_GT(sym_vars_.count(symbol), 0U) << "Not symbol is recorded for " << symbol;
+    std::vector<std::string> vars = sym_vars_[symbol];
+    for (const auto& it : vars) {
+      CHECK_GT(metadata_.count(it), 0U) << "Found not recorded constant variable: " << it;
+      ret.push_back(metadata_[it]);
+    }
+    return ret;
+  }
+
+  /*!
+   * \brief Initialize each imported module.
+   * \param symobl The symbol used for initializing a module. It is also used
+   * for runtime lookup.
+   *
+   * \note  A module could be like the following:
+   *  MetadataModuleNode (contains all the metadata)
+   *    - CSourceModule
+   *    - JSON runtime module
+   *
+   *  The initializer iterates through the imported modules and intilizes the
+   *  found module accordingly by passing the needed metadata into it.
+   */
+  void InitSubModule(const std::string& symbol) {
+    PackedFunc init(nullptr);
+    for (Module it : this->imports()) {
+      // Get the initialization function from the imported modules.
+      std::string init_name = "__init_" + symbol;
+      init = it.GetFunction(init_name, false);
+      if (init != nullptr) {
+        auto md = GetRequiredMetadata(symbol);
+        // Initialize the module with metadata.
+        init(md);
+        break;
+      }
+    }
+  }
+
+  void SaveToBinary(dmlc::Stream* stream) final {
+    std::vector<std::string> variables;
+    std::vector<NDArray> metadata;
+    for (const auto& it : metadata_) {
+      String var_name = it.first;
+      variables.push_back(var_name);
+      metadata.push_back(it.second);
+    }
+
+    // Save all variables in the function.
+    stream->Write(variables);
+    // Save all constant data.
+    uint64_t sz = static_cast<uint64_t>(metadata.size());
+    stream->Write(sz);
+    for (uint64_t i = 0; i < sz; i++) {
+      metadata[i].Save(stream);
+    }
+
+    // Save the symbol to list of required constant variables mapping
+    std::vector<std::string> symbols;
+    std::vector<std::vector<std::string>> const_vars;
+    for (const auto& it : sym_vars_) {
+      symbols.push_back(it.first);
+      const_vars.push_back(it.second);
+    }
+
+    stream->Write(symbols);
+    sz = static_cast<uint64_t>(sym_vars_.size());
+    stream->Write(sz);
+    for (uint64_t i = 0; i < sz; i++) {
+      stream->Write(const_vars[i]);
+    }
+  }
+
+  static Module LoadFromBinary(void* strm) {
+    dmlc::Stream* stream = static_cast<dmlc::Stream*>(strm);
+
+    // Load the variables.
+    std::vector<std::string> variables;
+    CHECK(stream->Read(&variables)) << "Loading variables failed";
+    uint64_t sz;
+    CHECK(stream->Read(&sz, sizeof(sz))) << "Loading medata size failed";
+    CHECK_EQ(static_cast<size_t>(sz), variables.size())
```

Comment:
I think we can keep this cast to make `size_t m_size = static_cast<size_t>(sz)` so that the rest loops in this function can all use `size_t i`.